File: Solutions/JavaScript/merge_sorted_array.js
/*
    [EASY]
    88. Merge Sorted Array

    Concepts:
    - array
    - two pointers

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 48.87 MB  [Beats 82.04%]
*/

/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */
var merge = function(nums1, m, nums2, n) {
    const nums3 = new Array(m+n);
    let i = 0, j = 0;

    while (i < m && j < n) {
        nums3[i+j] = (nums1[i] < nums2[j]) ? (nums1[i++]) : (nums2[j++]);
    }
    while (i < m) {
        nums3[n+i] = nums1[i++];
    }
    while (j < n) {
        nums3[m+j] = nums2[j++];
    }
    
    nums3.forEach((num, idx) => nums1[idx] = num);
};

File: Solutions/Rust/merge_sorted_array.rs
/*
    [EASY]
    88. Merge Sorted Array

    Concepts:
    - two pointers
    - sorting

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 2.33 MB   [Beats 1.09%]
*/

impl Solution {
    pub fn merge(nums1: &mut Vec<i32>, m: i32, nums2: &mut Vec<i32>, n: i32) {
        let (mut i, mut j) = (0 as usize, 0 as usize);
        while (
            i < (m + n).try_into().unwrap()
            && j < n.try_into().unwrap()
        ) {
            if nums1[i] > nums2[j] {
                nums1.splice(i..i, [nums2[j]].iter().cloned());
                j += 1;
            } else { i += 1; }
        }

        while j < n.try_into().unwrap() {
            nums1[m as usize + j] = nums2[j];
            j += 1;
        }

        nums1.drain((m + n) as usize..nums1.len());
    }
}